Dockerfile

```diff
@@ -12,32 +12,55 @@ RUN add-apt-repository ppa:nextgis/ppa
 
 # And now install apt dependencies, including a few of the heavy Python projects
 RUN apt-get update && apt-get install -y --no-install-recommends \
+    # Core requirements from travis.yml
     gdal-bin libgdal-dev libgdal20 libudunits2-0 \
-    python3 python3-gdal python3-setuptools python3-dev python3-numpy python3-netcdf4 \
+    # Extra python components, to speed things up
+    python3 python3-setuptools python3-dev python3-numpy python3-netcdf4 \
+    # Need pip to install more python packages later
     python3-pip \
+    # G++ because GDAL decided it needed compiling
+    g++ \
     && rm -rf /var/lib/apt/lists/*
 
-# Install psychopg2 as a special case, to quiet the warning message 
-RUN pip3 install --no-cache --no-binary :all: psycopg2
-
 # Get the code, and put it in /code
-ENV APPDIR=/code
+ENV APPDIR=/tmp/code
 RUN mkdir -p $APPDIR
 COPY . $APPDIR
 WORKDIR $APPDIR
 
 # Set the locale, this is required for some of the Python packages
 ENV LC_ALL C.UTF-8
+# And some GDAL variables
+ENV CPLUS_INCLUDE_PATH=/usr/include/gdal
+ENV C_INCLUDE_PATH=/usr/include/gdal
+
+# Install dependencies. First update pip
+RUN pip3 install --upgrade pip setuptools wheel \
+    && rm -rf $HOME/.cache/pip
 
-# Install dependencies
-RUN pip3 install '.[test,analytics,celery,s3]' --upgrade
-RUN pip3 install ./tests/drivers/fail_drivers --no-deps --upgrade
+# Install psychopg2 as a special case, to quiet the warning message 
+RUN pip3 install --no-cache --no-binary :all: psycopg2 \
+    && rm -rf $HOME/.cache/pip
+
+# Now use the setup.py file to identify dependencies
+RUN pip3 install '.[test,celery,s3]' --upgrade \
+    && rm -rf $HOME/.cache/pip
+RUN pip3 install ./tests/drivers/fail_drivers --no-deps --upgrade \
```

Comment:
We should not be installing test drivers, they will generate annoying warning when attempting to load them. This is by design to test failure paths in the driver loading code.